File: frontend/src/domain/agent.ts
import type { QuestionResult } from './knowledge'

export type AgentRunStatus = 'queued' | 'running' | 'completed' | 'failed' | 'canceled' | 'max_turns'

export type AgentRunSummary = {
  id: number
  question_id: number
  conversation_id: number
  turn_index: number | null
  status: AgentRunStatus
  stage: string
  current_turn: number
  max_turns: number
  tool_call_count: number
  stop_reason: string | null
  error: { code: string; message: string; retryable?: boolean } | null
  created_at: string
  completed_at: string | null
}

export type AgentActivityEvent = {
  sequence: number
  run_id: number
  turn: number
  type: string
  tool: string | null
  label: string | null
  status: 'started' | 'completed' | 'failed' | 'terminal' | null
  query_preview: string | null
  node_labels: string[]
  result_count: number | null
  error_code: string | null
  created_at: string
}

export type AgentWebSource = {
  citation_key: string
  url: string
  title: string | null
  publisher: string | null
  rank: number
}

export type AgentRunPayload = {
  run: AgentRunSummary
  result: QuestionResult | null
  web_sources: AgentWebSource[]
}

File: backend/app/schemas/agent.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class AgentRunCreate(BaseModel):
    question: str = Field(min_length=2, max_length=1000)
    conversation_id: int | None = Field(default=None, ge=1)


class AgentRunSummary(BaseModel):
    id: int
    question_id: int
    conversation_id: int
    turn_index: int | None
    status: str
    stage: str
    current_turn: int
    max_turns: int
    tool_call_count: int
    stop_reason: str | None = None
    error: dict[str, Any] | None = None
    created_at: datetime
    completed_at: datetime | None = None


class AgentActivityEvent(BaseModel):
    sequence: int
    run_id: int
    turn: int
    type: str
    tool: str | None = None
    label: str | None = None
    status: Literal["started", "completed", "failed", "terminal"] | None = None
    query_preview: str | None = None
    node_labels: list[str] = Field(default_factory=list)
    result_count: int | None = None
    error_code: str | None = None
    created_at: datetime
